### PR TITLE
Extend Supervisor

### PR DESCRIPTION
**Pull request by SupaHam**
Supervisor is a Bukkit plugin that arms admins with a tool of data gathering for sharing with people such as developers in need of tracking down reported issues with their plugins. Essentials has a lot to gain from Supervisor in regards to issue squashing.

In this PR the EssentialsReportContext class is created which extends ReportContext and ReportContextEntry (for convenience) in order to provide details relevant to issues such as the config.yml and online user files.

Currently the following data is being logged:

| key | type | description |
| --- | --- | --- |
| users_count | integer | Total unique users that Essentials has information about. |
| uuidmap_count | integer | Total number of entries in the usermap.csv file. |

Files:

| filename | report level | description |
| --- | --- | --- |
| config.yml | >200 | A copy of the config.yml (from disk) |
| userdata/<uuid/name>.yml | >=400 | Every online user at the time of report has their file copied (from disk). |

Here's an example of a _brief_ report generated by Supervisor: https://gist.github.com/anonymous/7992db7eec6a6940a966

Here's an example of a more verbose (normal) report generated by Supervisor: https://gist.github.com/anonymous/f5afcb5452591a82e766

Default values may be tweaked for convenience.

This PR does **NOT** make Essentials depend on Supervisor, it is merely an optional extension.

### CODE REVIEW

**Comment by vemacs** [comment]
I think this should be pulled to figure out the cause of the UUIDMap wipe (and remove the Mojang lookup workaround).

**Comment by SupaHam** [comment]
How exactly can we track down the UUIDMap wipe with the current code? All it does currently is count of unique user files and entries in the `names` map (in memory, thus usermap.csv file).

File: Essentials/src/com/earth2me/essentials/EssentialsPlayerListener.java [hunk 1]
```diff
@@ -384,25 +384,25 @@ public void onPlayerCommandPreprocess(final PlayerCommandPreprocessEvent event)
...
+                    update = false;
+                case "vanish":
+                    broadcast = false;
             }
         }
+        if (update) {
+            final User user = ess.getUser(player);
+            user.updateActivity(broadcast);
+        }
     }
```
Comment (drtshock) [inline Essentials/src/com/earth2me/essentials/EssentialsPlayerListener.java:406]: Why are changes in this class included in this PR? They seem unrelated.

Comment (SupaHam) [inline Essentials/src/com/earth2me/essentials/EssentialsPlayerListener.java:406]: I accidently pushed a later commit, but this commit is already live.

**Comment by SupaHam** [comment]
I'd really like to see this used and recommended to users. Overall, I think it's a lot more convenient for both the developers and the reporters.

The reporters tend to not provide information off the bat, if we give them a simple command to type (`/sv`) and they just post the URL within their issue I think we could get things done a lot faster.

This also adds a lot more validity to report data in the sense that we can exactly pin point related issues to things such as plugin versions and configurations.

**Comment by mdcfe** [comment]
+1, support for this would be incredibly useful in gathering essential details for supporting people who are experiencing issues.

**Comment by Zarthus** [comment]
I've had an intensive, extensive, exhaustive, and inflammatory discussion with drtshock and we have come to the conclusion that aside from some minor issues (like the merge conflicts!) you've got my permission to merge this.

Thanks for keeping essentials on top of the food chain.

File: Essentials/pom.xml [hunk 2]
```diff
@@ -128,6 +128,12 @@
             <version>2.0.1</version>
             <scope>compile</scope>
         </dependency>
+        <dependency>
+            <groupId>com.supaham.supervisor</groupId>
```
Comment (Zarthus) [inline Essentials/pom.xml:132]: I like this part especially.

**Review by Zarthus — APPROVED** [review]

**Comment by SupaHam** [comment]
Merged.
